### ISSUE TITLE
Generalize spline tax functions to 2D

### ISSUE THREAD

**Issue by jdebacker** [PSLmodels/OG-Core#828]
This issue is to discuss the extrapolation of the spline tax functions introduced in PR #819 to two dimensions, where those two dimensions are capital and labor income.

The key is to continue to implement the constraints that ensure the functions are monotonic.

cc @prrathi @rickecon

**Comment by rickecon** [comment]
The code for our one-dimensional monotone smoothing spline is in [`txfunc.py`](https://github.com/PSLmodels/OG-Core/blob/master/ogcore/txfunc.py) lines 669-690, 1635-1780 (mainly lines 1646-1694). This approach was taken from this site and these citations [Eilers (2005, 2003) and Eilers and Marx (1996)](https://pslmodels.github.io/OG-Core/content/OGCore_references.html#id15). This approach is a smoothing spline with a penalty on any  nonmonotonicities (negative slope).

**Comment by prrathi** [comment]
@rickecon Read through the code, from my understanding the iterations are penalizing nonmonotonicities by having large opposite-sign values in the matrix when solving for coefficients, so as to force them to the same value, is that right? A few questions, the `D3` seemed to be there to smooth coefficients, is there a reason that's chosen over other numbers? Also is it fine if the solution is slightly not monotonic, it seems in the test code that there was very small nonmonotonicity at the end of the coefficients obtained? Also was curious what the identity matrix was added in for, as it seems to be on a much smaller scale than the other values in the matrix. I'll take a look at the papers if that gives more reference.
Lastly for the 2d monotone, would the idea be a similar approach but two sets of systems for the two dimensions? Thanks for the help!

**Comment by prrathi** [comment]
@rickecon one more follow-up question, in the `mono_interp` function of `txfunc.py` even though `y_cstr` is monotonic does that necessarily guarantee the cubic interpolation is monotonic?

**Comment by jdebacker** [comment]
the [`cm.spline` function](https://search.r-project.org/CRAN/refmans/demography/html/cm.spline.html) in R seems relevant here.

**Comment by rickecon** [comment]
Some other references:
* Pya, Natalya and Simon N. Wood, "[Shape constrained additive models](https://link.springer.com/article/10.1007/s11222-013-9448-7)", Statistics and Computing, vol. 25 pp. 543-559 (2015).
* Another [R reference to forcing monotonicity with smoothing splines](https://stats.stackexchange.com/questions/197509/how-to-smooth-data-and-force-monotonicity). But examples are in 1D.
* Python's machine learning stack `scikit-learn` has a documentation page on "[Monotonic constraints](https://scikit-learn.org/stable/auto_examples/ensemble/plot_monotonic_constraints.html)". The example is fitting a Histogram-based Gradient Boost Regression Tree to data with monotonic constraints. I think this method could be applied to 2D data. A related example is [here](https://scikit-learn.ru/example/monotonic-constraints/).

cc: @jdebacker @prrathi

**Comment by prrathi** [comment]
@jdebacker @rickecon would you happen to have a pdf version of Unimodal smoothing by Eilers? I don't think I have institutional access to the paper

**Comment by prrathi** [comment]
@jdebacker @rickecon is monotonicity the only requirement here or does smoothness/continuity of derivatives matter

**Comment by rickecon** [comment]
@prrathi monotonicity and C2 (continuity in the levels and in the first derivative) I don't think we need continuity in the second derivative, but I don't know. It probably depends on the optimizer. But I think a property of generalized additive models is that if the kernel is C2, then the final predictor is C2.

**Comment by prrathi** [comment]
@rickecon making some progress here with [pyGAM](https://pygam.readthedocs.io/en/latest/index.html) that allows monotonic constraints and smoothing up to the second derivative. I need to test it out with different data- would you have a good source for tax data like plotted [here](https://pslmodels.github.io/OG-Core/content/theory/government.html#figtaxcalc3dvspred)?

**Comment by jdebacker** [comment]
@prrathi You could work with the micro data (produced from a Current Population Survey data file) that is in [`micro_data_for_tests.pkl`](https://github.com/PSLmodels/OG-Core/blob/master/tests/test_io_data/micro_data_dict_for_tests.pkl)

**Comment by prrathi** [comment]
@rickecon @jdebacker able to fit the tax data using pyGAM. I first bin the data along both axes- labor income and capital income- (wrote a general function that should apply to any number of dimensions) and then fit the model using the weighted results. There are parameters/configurations to test around with like number and degree of splines, I've attached some pictures, the following three are different angles of same result for 100 by 100 bins with 20 splines for each labor and capital income:
![angle3](https://user-images.githubusercontent.com/53785742/209419326-a2007ac1-fb09-448a-8f9d-3e58c274e8dd.png)
![angle2](https://user-images.githubusercontent.com/53785742/209419328-e1f4844c-19fa-42c4-9097-31d39f9131c4.png)
![origAngle](https://user-images.githubusercontent.com/53785742/209419330-2d14ce1f-2c92-4764-ab8e-f157e6790959.png)

These three are for 100 by 100 bins with 100 splines for each labor and capital income:
![2angle3](https://user-images.githubusercontent.com/53785742/209419357-8935f1c3-6d4f-4ba9-b0ee-af17d554a830.png)
![2angle2](https://user-images.githubusercontent.com/53785742/209419358-df577f0d-2648-4650-9c9a-8fc51f152723.png)
![2origangle](https://user-images.githubusercontent.com/53785742/209419360-b776e6d1-e9c2-41c0-a1b6-2e19e25abcac.png)

The data for the plots is only from the middle 50% of both labor and capital incomes to make trends more clear, but the models were trained on all data. For 20 splines fit mean deviation for middle 50% of both incomes is 0.08669 (units of etr), for 100 splines mean deviation for middle 50% is 0.08517, excluded some of the really big outliers (especially on the upper end) for this comparison
